x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/action/GetCheckpointNodeAction.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.transform.action;
+
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.action.ActionType;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.index.shard.ShardId;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.Set;
+
+public class GetCheckpointNodeAction extends ActionType<GetCheckpointNodeAction.Response> {
+
+    public static final GetCheckpointNodeAction INSTANCE = new GetCheckpointNodeAction();
+
+    // note: this is an internal action
+    public static final String NAME = "cluster:internal/transform/checkpoint[n]";
```

Comment:
I find it weird that this action is a clusterAction but the request still target indices (shards). I really think this should be an indexAction as well in the format of `indices:internal/transform/checkpoint[n]`. An example in the existing code and also similar to the scenario here is the FieldCaps request which is an indicesRequest:
https://github.com/elastic/elasticsearch/blob/2982591e6d8ee20fa5d7d78da1d0b456bbe056a4/server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesNodeRequest.java#L27

and the action name is `indices:data/read/field_caps[n]`:
https://github.com/elastic/elasticsearch/blob/2982591e6d8ee20fa5d7d78da1d0b456bbe056a4/server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java#L52

Renaming it to be `indices:internal/transform/checkpoint[n]` also means it will be covered by the same `view_index_metadata` index privilege. So I suggest we rename the action to `indices:...` and make the associated request implements `IndicesRequest`.

Reply:
I stumbled upon the same thing. With `IndicesRequest` we carry the index name which is actually not necessary after resolving the shards, that's the reason why I kept it a cluster action. However it probably doesn't matter to send a few bytes more.